File: src/syntax/bibtex/ast.ts
import { Position, Range } from 'vscode-languageserver';
import { SyntaxNode } from '../syntaxNode';
import { BibtexToken } from './lexer';

export enum BibtexSyntaxKind {
  Document,
  Comment,
  Preamble,
  String,
  Entry,
  Field,
  Word,
  Command,
  QuotedContent,
  BracedContent,
  Concat,
}

export type BibtexSyntaxNode =
  | BibtexDocumentSyntax
  | BibtexCommentSyntax
  | BibtexDeclarationSyntax
  | BibtexFieldSyntax
  | BibtexContentSyntax;

export class BibtexDocumentSyntax extends SyntaxNode {
  public readonly kind: BibtexSyntaxKind.Document;
  public readonly range: Range;

  constructor(public readonly children: BibtexDocumentItemSyntax[]) {
    super();
    this.kind = BibtexSyntaxKind.Document;
    this.range =
      children.length === 0
        ? { start: { line: 0, character: 0 }, end: { line: 0, character: 0 } }
        : { start: children[0].start, end: children[children.length - 1].end };
  }
}

export type BibtexDocumentItemSyntax =
  | BibtexCommentSyntax
  | BibtexDeclarationSyntax;

export class BibtexCommentSyntax extends SyntaxNode {
  public readonly kind: BibtexSyntaxKind.Comment;
  public readonly range: Range;

  constructor(public readonly token: BibtexToken) {
    super();
    this.kind = BibtexSyntaxKind.Comment;
    this.range = token.range;
  }
}

export type BibtexDeclarationSyntax =
  | BibtexPreambleSyntax
  | BibtexStringSyntax
  | BibtexEntrySyntax;

export class BibtexPreambleSyntax extends SyntaxNode {
  public kind: BibtexSyntaxKind.Preamble;
  public range: Range;

  constructor(
    public readonly type: BibtexToken,
    public readonly left: BibtexToken | undefined,
    public readonly content: BibtexContentSyntax | undefined,
    public readonly right: BibtexToken | undefined,
  ) {
    super();
    this.kind = BibtexSyntaxKind.Preamble;
    let end: Position;
    if (right !== undefined) {
      end = type.end;
    } else if (content !== undefined) {
      end = content.end;
    } else if (left !== undefined) {
      end = left.end;
    } else {
      end = type.end;
    }
    this.range = { start: type.start, end };
  }
}

export class BibtexStringSyntax extends SyntaxNode {
  public readonly kind: BibtexSyntaxKind.String;
  public readonly range: Range;

  constructor(
    public readonly type: BibtexToken,
    public readonly left: BibtexToken | undefined,
    public readonly name: BibtexToken | undefined,
    public readonly assign: BibtexToken | undefined,
    public readonly value: BibtexContentSyntax | undefined,
    public readonly right: BibtexToken | undefined,
  ) {
    super();
    this.kind = BibtexSyntaxKind.String;
    let end: Position;
    if (right !== undefined) {
      end = right.end;
    } else if (value !== undefined) {
      end = value.end;
    } else if (assign !== undefined) {
      end = assign.end;
    } else if (name !== undefined) {
      end = name.end;
    } else if (left !== undefined) {
      end = left.end;
    } else {
      end = type.end;
    }
    this.range = { start: type.start, end };
  }
}

export class BibtexEntrySyntax extends SyntaxNode {
  public readonly kind: BibtexSyntaxKind.Entry;
  public readonly range: Range;

  constructor(
    public readonly type: BibtexToken,
    public readonly left: BibtexToken | undefined,
    public readonly name: BibtexToken | undefined,
    public readonly comma: BibtexToken | undefined,
    public readonly fields: BibtexFieldSyntax[],
    public readonly right: BibtexToken | undefined,
  ) {
    super();
    this.kind = BibtexSyntaxKind.Entry;
    let end: Position;
    if (right !== undefined) {
      end = right.end;
    } else if (fields.length > 0) {
      end = fields[fields.length - 1].end;
    } else if (comma !== undefined) {
      end = comma.end;
    } else if (name !== undefined) {
      end = name.end;
    } else if (left !== undefined) {
      end = left.end;
    } else {
      end = type.end;
    }
    this.range = { start: type.start, end };
  }
}

export class BibtexFieldSyntax extends SyntaxNode {
  public readonly kind: BibtexSyntaxKind.Field;
  public readonly range: Range;

  constructor(
    public readonly name: BibtexToken,
    public readonly assign: BibtexToken | undefined,
    public readonly content: BibtexContentSyntax | undefined,
    public readonly comma: BibtexToken | undefined,
  ) {
    super();
    this.kind = BibtexSyntaxKind.Field;
    let end: Position;
    if (comma !== undefined) {
      end = comma.end;
    } else if (content !== undefined) {
      end = content.end;
    } else if (assign !== undefined) {
      end = assign.end;
    } else {
      end = name.end;
    }
    this.range = { start: name.start, end };
  }
}

type BibtexContentSyntax =
  | BibtexWordSyntax
  | BibtexCommandSyntax
  | BibtexQuotedContentSyntax
  | BibtexBracedContentSyntax
  | BibtexConcatSyntax;

export class BibtexWordSyntax extends SyntaxNode {
  public readonly kind: BibtexSyntaxKind.Word;
  public readonly range: Range;

  constructor(public readonly token: BibtexToken) {
    super();
    this.kind = BibtexSyntaxKind.Word;
    this.range = token.range;
  }
}

export class BibtexCommandSyntax extends SyntaxNode {
  public readonly kind: BibtexSyntaxKind.Command;
  public readonly range: Range;

  constructor(public readonly token: BibtexToken) {
    super();
    this.kind = BibtexSyntaxKind.Command;
    this.range = token.range;
  }
}

export class BibtexQuotedContentSyntax extends SyntaxNode {
  public readonly kind: BibtexSyntaxKind.QuotedContent;
  public readonly range: Range;

  constructor(
    public readonly left: BibtexToken,
    public readonly children: BibtexContentSyntax[],
    public readonly right: BibtexToken | undefined,
  ) {
    super();
    this.kind = BibtexSyntaxKind.QuotedContent;
    let end: Position;
    if (right !== undefined) {
      end = right.end;
    } else if (children.length > 0) {
      end = children[children.length - 1].end;
    } else {
      end = left.end;
    }
    this.range = { start: left.start, end };
  }
}

export class BibtexBracedContentSyntax extends SyntaxNode {
  public readonly kind: BibtexSyntaxKind.BracedContent;
  public readonly range: Range;

  constructor(
    public readonly left: BibtexToken,
    public readonly children: BibtexContentSyntax[],
    public readonly right: BibtexToken | undefined,
  ) {
    super();
    this.kind = BibtexSyntaxKind.BracedContent;
    let end: Position;
    if (right !== undefined) {
      end = right.end;
    } else if (children.length > 0) {
      end = children[children.length - 1].end;
    } else {
      end = left.end;
    }
    this.range = { start: left.start, end };
  }
}

export class BibtexConcatSyntax extends SyntaxNode {
  public readonly kind: BibtexSyntaxKind.Concat;
  public readonly range: Range;

  constructor(
    public readonly left: BibtexContentSyntax,
    public readonly operator: BibtexToken,
    public readonly right: BibtexContentSyntax | undefined,
  ) {
    super();
    this.kind = BibtexSyntaxKind.Concat;
    const end = right === undefined ? operator.end : right.end;
    this.range = { start: left.start, end };
  }
}

export function descendants(root: BibtexSyntaxNode) {
  const results: BibtexSyntaxNode[] = [];
  function visit(node: BibtexSyntaxNode) {
    results.push(node);
    switch (node.kind) {
      case BibtexSyntaxKind.Document:
        node.children.forEach(visit);
        break;
      case BibtexSyntaxKind.Comment:
        break;
      case BibtexSyntaxKind.Preamble:
        if (node.content !== undefined) {
          visit(node.content);
        }
        break;
      case BibtexSyntaxKind.String:
        if (node.value !== undefined) {
          visit(node.value);
        }
        break;
      case BibtexSyntaxKind.Entry:
        node.fields.forEach(visit);
        break;
      case BibtexSyntaxKind.Field:
        if (node.content !== undefined) {
          visit(node.content);
        }
        break;
      case BibtexSyntaxKind.Word:
        break;
      case BibtexSyntaxKind.Command:
        break;
      case BibtexSyntaxKind.QuotedContent:
        node.children.forEach(visit);
        break;
      case BibtexSyntaxKind.BracedContent:
        node.children.forEach(visit);
        break;
      case BibtexSyntaxKind.Concat:
        visit(node.left);
        if (node.right !== undefined) {
          visit(node.right);
        }
        break;
    }
  }
  visit(root);
  return results;
}

File: crates/texlab_syntax/src/bibtex/ast.rs
use crate::{Span, SyntaxNode};
use itertools::Itertools;
use petgraph::graph::{Graph, NodeIndex};
use serde::{Deserialize, Serialize};
use std::fmt;
use texlab_protocol::{Position, Range, RangeExt};

#[derive(Debug, PartialEq, Eq, Clone, Copy, Serialize, Deserialize)]
pub enum TokenKind {
    PreambleKind,
    StringKind,
    EntryKind,
    Word,
    Command,
    Assign,
    Comma,
    Concat,
    Quote,
    BeginBrace,
    EndBrace,
    BeginParen,
    EndParen,
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
pub struct Token {
    pub span: Span,
    pub kind: TokenKind,
}

impl SyntaxNode for Token {
    fn range(&self) -> Range {
        self.span.range()
    }
}

impl Token {
    pub fn new(span: Span, kind: TokenKind) -> Self {
        Self { span, kind }
    }

    pub fn text(&self) -> &str {
        &self.span.text
    }
}

#[derive(PartialEq, Eq, Clone, Copy, Serialize, Deserialize)]
pub struct Root {
    pub range: Range,
}

impl SyntaxNode for Root {
    fn range(&self) -> Range {
        self.range
    }
}

impl fmt::Debug for Root {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Root")
    }
}

#[derive(PartialEq, Eq, Clone, Serialize, Deserialize)]
pub struct Comment {
    pub token: Token,
}

impl SyntaxNode for Comment {
    fn range(&self) -> Range {
        self.token.range()
    }
}

impl fmt::Debug for Comment {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Comment({})", self.token.text())
    }
}

#[derive(PartialEq, Eq, Clone, Serialize, Deserialize)]
pub struct Preamble {
    pub range: Range,
    pub ty: Token,
    pub left: Option<Token>,
    pub right: Option<Token>,
}

impl SyntaxNode for Preamble {
    fn range(&self) -> Range {
        self.range
    }
}

impl fmt::Debug for Preamble {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Preamble")
    }
}

#[derive(PartialEq, Eq, Clone, Serialize, Deserialize)]
pub struct String {
    pub range: Range,
    pub ty: Token,
    pub left: Option<Token>,
    pub name: Option<Token>,
    pub assign: Option<Token>,
    pub right: Option<Token>,
}

impl SyntaxNode for String {
    fn range(&self) -> Range {
        self.range
    }
}

impl fmt::Debug for String {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "String({:?})", self.name.as_ref().map(Token::text))
    }
}

#[derive(PartialEq, Eq, Clone, Serialize, Deserialize)]
pub struct Entry {
    pub range: Range,
    pub ty: Token,
    pub left: Option<Token>,
    pub key: Option<Token>,
    pub comma: Option<Token>,
    pub right: Option<Token>,
}

impl SyntaxNode for Entry {
    fn range(&self) -> Range {
        self.range
    }
}

impl fmt::Debug for Entry {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Entry({:?})", self.key.as_ref().map(Token::text))
    }
}

impl Entry {
    pub fn is_comment(&self) -> bool {
        self.ty.text().to_lowercase() == "@comment"
    }
}

#[derive(PartialEq, Eq, Clone, Serialize, Deserialize)]
pub struct Field {
    pub range: Range,
    pub name: Token,
    pub assign: Option<Token>,
    pub comma: Option<Token>,
}

impl SyntaxNode for Field {
    fn range(&self) -> Range {
        self.range
    }
}

impl fmt::Debug for Field {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Field({})", self.name.text())
    }
}

#[derive(PartialEq, Eq, Clone, Serialize, Deserialize)]
pub struct Word {
    pub token: Token,
}

impl SyntaxNode for Word {
    fn range(&self) -> Range {
        self.token.range()
    }
}

impl fmt::Debug for Word {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Word({})", self.token.text())
    }
}

#[derive(PartialEq, Eq, Clone, Serialize, Deserialize)]
pub struct Command {
    pub token: Token,
}

impl SyntaxNode for Command {
    fn range(&self) -> Range {
        self.token.range()
    }
}

impl fmt::Debug for Command {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Command({})", self.token.text())
    }
}

#[derive(PartialEq, Eq, Clone, Serialize, Deserialize)]
pub struct QuotedContent {
    pub range: Range,
    pub left: Token,
    pub right: Option<Token>,
}

impl SyntaxNode for QuotedContent {
    fn range(&self) -> Range {
        self.range
    }
}

impl fmt::Debug for QuotedContent {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "QuotedContent")
    }
}

#[derive(PartialEq, Eq, Clone, Serialize, Deserialize)]
pub struct BracedContent {
    pub range: Range,
    pub left: Token,
    pub right: Option<Token>,
}

impl SyntaxNode for BracedContent {
    fn range(&self) -> Range {
        self.range
    }
}

impl fmt::Debug for BracedContent {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "BracedContent")
    }
}

#[derive(PartialEq, Eq, Clone, Serialize, Deserialize)]
pub struct Concat {
    pub range: Range,
    pub operator: Token,
}

impl SyntaxNode for Concat {
    fn range(&self) -> Range {
        self.range
    }
}

impl fmt::Debug for Concat {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "Concat")
    }
}

#[derive(Debug, PartialEq, Eq, Clone, Serialize, Deserialize)]
pub enum Node {
    Root(Root),
    Comment(Comment),
    Preamble(Box<Preamble>),
    String(Box<String>),
    Entry(Box<Entry>),
    Field(Box<Field>),
    Word(Word),
    Command(Command),
    QuotedContent(QuotedContent),
    BracedContent(BracedContent),
    Concat(Concat),
}

impl SyntaxNode for Node {
    fn range(&self) -> Range {
        match self {
            Self::Root(root) => root.range(),
            Self::Comment(comment) => comment.range(),
            Self::Preamble(preamble) => preamble.range(),
            Self::String(string) => string.range(),
            Self::Entry(entry) => entry.range(),
            Self::Field(field) => field.range(),
            Self::Word(word) => word.range(),
            Self::Command(cmd) => cmd.range(),
            Self::QuotedContent(content) => content.range(),
            Self::BracedContent(content) => content.range(),
            Self::Concat(concat) => concat.range(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Tree {
    pub graph: Graph<Node, ()>,
    pub root: NodeIndex,
}

impl Tree {
    pub fn children(&self, parent: NodeIndex) -> impl Iterator<Item = NodeIndex> {
        self.graph
            .neighbors(parent)
            .sorted_by_key(|child| self.graph[*child].start())
    }

    pub fn has_children(&self, parent: NodeIndex) -> bool {
        self.children(parent).next().is_some()
    }

    pub fn walk<'a, V: Visitor<'a>>(&'a self, visitor: &mut V, parent: NodeIndex) {
        for child in self.children(parent) {
            visitor.visit(self, child);
        }
    }

    pub fn find(&self, pos: Position) -> Vec<NodeIndex> {
        let mut finder = Finder::new(pos);
        finder.visit(self, self.root);
        finder.results
    }

    pub fn as_preamble(&self, node: NodeIndex) -> Option<&Preamble> {
        if let Node::Preamble(preamble) = &self.graph[node] {
            Some(preamble)
        } else {
            None
        }
    }

    pub fn as_string(&self, node: NodeIndex) -> Option<&String> {
        if let Node::String(string) = &self.graph[node] {
            Some(string)
        } else {
            None
        }
    }

    pub fn as_entry(&self, node: NodeIndex) -> Option<&Entry> {
        if let Node::Entry(entry) = &self.graph[node] {
            Some(entry)
        } else {
            None
        }
    }

    pub fn as_field(&self, node: NodeIndex) -> Option<&Field> {
        if let Node::Field(field) = &self.graph[node] {
            Some(field)
        } else {
            None
        }
    }

    pub fn as_command(&self, node: NodeIndex) -> Option<&Command> {
        if let Node::Command(cmd) = &self.graph[node] {
            Some(cmd)
        } else {
            None
        }
    }

    pub fn as_word(&self, node: NodeIndex) -> Option<&Word> {
        if let Node::Word(word) = &self.graph[node] {
            Some(word)
        } else {
            None
        }
    }

    pub fn entry_by_key(&self, key: &str) -> Option<NodeIndex> {
        for node in self.children(self.root) {
            if let Some(entry) = self.as_entry(node) {
                if entry.key.as_ref().map(Token::text) == Some(key) {
                    return Some(node);
                }
            }
        }
        None
    }

    pub fn field_by_name(&self, parent: NodeIndex, name: &str) -> Option<NodeIndex> {
        let name = name.to_lowercase();
        self.as_entry(parent)?;
        for node in self.children(parent) {
            if let Some(field) = self.as_field(node) {
                if field.name.text() == name {
                    return Some(node);
                }
            }
        }
        None
    }

    pub fn crossref(&self, entry: NodeIndex) -> Option<NodeIndex> {
        let field = self.field_by_name(entry, "crossref")?;
        let content = self.children(field).next()?;
        let key = self.as_word(self.children(content).next()?)?;
        self.entry_by_key(key.token.text())
    }
}

pub trait Visitor<'a> {
    fn visit(&mut self, tree: &'a Tree, node: NodeIndex);
}

#[derive(Debug)]
struct Finder {
    position: Position,
    results: Vec<NodeIndex>,
}

impl Finder {
    fn new(position: Position) -> Self {
        Self {
            position,
            results: Vec::new(),
        }
    }
}

impl<'a> Visitor<'a> for Finder {
    fn visit(&mut self, tree: &'a Tree, node: NodeIndex) {
        if tree.graph[node].range().contains(self.position) {
            self.results.push(node);
            tree.walk(self, node);
        }
    }
}
